forum/qa-include/pages/ask.php

```diff
@@ -1,290 +1,302 @@
+		qa_db_popular_tags_selectspec(0, QA_DB_RETRIEVE_COMPLETE_TAGS)
+	);
+
+	if (!isset($categories[$in['categoryid']]))
+		$in['categoryid']=null;
+
+	if (@$followanswer['basetype']!='A')
+		$followanswer=null;
+
+
+//	Check for permission error
+
+	$permiterror=qa_user_maximum_permit_error('permit_post_q', QA_LIMIT_QUESTIONS);
+
+	if ($permiterror) {
+		$qa_content=qa_content_prepare();
+
+		// The 'approve', 'login', 'confirm', 'limit', 'userblock', 'ipblock' permission errors are reported to the user here
+		// The other option ('level') prevents the menu option being shown, in qa_content_prepare(...)
+
+		switch ($permiterror) {
+			case 'login':
+				$qa_content['error']=qa_insert_login_links(qa_lang_html('question/ask_must_login'), qa_request(), isset($followpostid) ? array('follow' => $followpostid) : null);
+				break;
+
+			case 'confirm':
+				$qa_content['error']=qa_insert_login_links(qa_lang_html('question/ask_must_confirm'), qa_request(), isset($followpostid) ? array('follow' => $followpostid) : null);
+				break;
+
+			case 'limit':
+				$qa_content['error']=qa_lang_html('question/ask_limit');
+				break;
+
+			case 'approve':
+				$qa_content['error']=qa_lang_html('question/ask_must_be_approved');
+				break;
+
+			default:
+				$qa_content['error']=qa_lang_html('users/no_permission');
+				break;
+		}
+
+		return $qa_content;
+	}
+
+
+//	Process input
+
+	$captchareason=qa_user_captcha_reason();
+
+	$in['title']=qa_get_post_title('title'); // allow title and tags to be posted by an external form
+	$in['extra']=qa_opt('extra_field_active') ? qa_post_text('extra') : null;
+	if (qa_using_tags())
+		$in['tags']=qa_get_tags_field_value('tags');
+
+	if (qa_clicked('doask')) {
+		require_once QA_INCLUDE_DIR.'app/post-create.php';
+		require_once QA_INCLUDE_DIR.'util/string.php';
+
+		$categoryids=array_keys(qa_category_path($categories, @$in['categoryid']));
+		$userlevel=qa_user_level_for_categories($categoryids);
+
+		$in['name']=qa_post_text('name');
+		$in['notify'] = strlen(qa_post_text('notify')) > 0;
+		$in['email']=qa_post_text('email');
+		$in['queued'] = qa_user_moderation_reason($userlevel) !== false;
+
+		qa_get_post_content('editor', 'content', $in['editor'], $in['content'], $in['format'], $in['text']);
+
+		$errors=array();
+
+		if (!qa_check_form_security_code('ask', qa_post_text('code')))
+			$errors['page']=qa_lang_html('misc/form_security_again');
+
+		else {
+			$filtermodules=qa_load_modules_with('filter', 'filter_question');
+			foreach ($filtermodules as $filtermodule) {
+				$oldin=$in;
+				$filtermodule->filter_question($in, $errors, null);
+				qa_update_post_text($in, $oldin);
+			}
+
+            $categoriesEnabled = qa_using_categories() && count($categories) > 0;
+            $noSubcategory = qa_opt('allow_no_sub_category');
+            $mainCategory = qa_post_text('category_1');
+
+            $subcategories = array_filter($categories, function ($category) use ($mainCategory) {
```

Comment:
Można zmienić nazwę tej zmiennej na coś w rodzaju `$mainSubCategories`? W funkcji jest sprawdzanie, czy parent subkategorii, to główna kategoria - więc już zmienna by więcej o sobie mówiła. Po drugie, pod spodem do tej samej zmiennej jest przypisywane coś innego, więc ta zmienna w następujących po sobie operacjach ma przypisane chyba inne wartości.

Reply:
Można, ale ta pierwsza ma main, ponieważ to główna wybrana kategoria, a druga ma subcategories, ponieważ zawiera podkategorie tej głównej. Wydało mi się więc logiczne nazwanie tego subcategories. To co ma to najpierw wybiera wszystkie dane podkategorii w array_filter, a później już wyciąga same id przez array_column

Reply:
> To co ma to najpierw wybiera wszystkie dane podkategorii w array_filter, a później już wyciąga same id przez array_column

Jak dla mnie z tego co piszesz wynika, że jednak wartość przypisana do zmiennej `$subcategories` w następujących po sobie miejscach jest jednak różna - najpierw wszystkie dane podkategorii, a później tylko id-ki. Dlatego proponuję, aby to były osobne zmienne; druga - skoro przechowuje tylko id-ki podkategorii - może się nazywać np. `$subcategoriesIDs`.

Początkowo chciałem zaproponować chainowanie tych dwóch operacji, ale tu się raczej nie da tego zrobić, bo funkcje `array_filter` i `array_column` nie są metodami tablicy a jakimiś luźnymi funkcjami, które jako argument przyjmują tę tablicę. Można by zrobić `array_column(array_filter( $subcategories, fn() ), fn() )`, ale to raczej pogorszy czytelność kodu. Dlatego skłoniłem się ku drugiej opcji - oddzieleniu tych zmiennych poprzez zmianę ich nazwy, bo mają przypisane inne wartości.

Reply:
No tak, nigdzie nie napisałem że jest identyczna :P Tyle że najpierw ma wszystkie dane podkategorii, a później same idki tych samych podkategorii, a całe dane podkategorii nie są nigdzie potem używane. Stąd po prostu napisałem wartość zmiennej. Ale ok, to drobiazg, zmieniłem.